docs: macOS development environment setup (#4054)
inveniosoftware/invenio: docs/getting-started/development-environment.rst

Developer environment guide
===========================

The following is a guide to help you prepare your system for developing with
Invenio. A proper development environment setup can save a lot of time and
frustrations. Note that the following guide, is not meant as a guide for
setting up servers.

System setup
------------
We support development on macOS and Linux-based systems (development on Windows
is not supported).

macOS
~~~~~
We recommend that you install libraries and tools using the
`Homebrew <https://brew.sh>`_ package manager. Homebrew allows you to install
two types of packages - cask and normal.

**Cask packages**

> Cask packages are usually UI applications that ends up en your Applications

Reviewer: 🇪🇸 
```suggestion
Cask packages are usually UI applications that end up in your Applications
```

folder. You can install cask packages using the command:

.. code-block:: console

    $ brew cask install <packages>

Here are some recommended cask packages:

.. code-block:: console

    discord            # app that we use for chat among developers
    docker             # docker setup for macOS
    google-chrome      # needed for some end-to-end tests
    iterm2             # a better terminal than Terminal.app

>     postico            # UI for PostgreSQL

Reviewer: Not to polute, so it's ok as it is. I use `DB browser for SQLite` and it works pretty fine for sqlite. https://sqlitebrowser.org

Author: Let's add it :-) Idea is that people can select what they want. Probably I should add only docker is really required
``db-browser-for-sqlite``

    spectacle          # organise windows with keyboard shortcuts
    visual-studio-code # a text editor used by many Invenio developers

**Normal packages**

> Normal packages are usually command line tools/libraries that ends up in

Reviewer: ```suggestion
Normal packages are usually command line tools/libraries that end up in
```

/usr/local and that you use from the CLI. You can install normal packages
using the command:

.. code-block:: console

    $ brew install <packages>

The following packages are libraries that are likely to be needed during
Invenio development in order to install certain Python packages:

.. code-block:: console

    cairo
    freetype
    geoip
    gettext
    glib
    imagemagick
    jpeg
    libffi
    libmemcached
    libpng
    libtiff
    libxml2
    libxslt
    readline
    xz
    zlib

> Following is CLI tools that are useful during development:

Reviewer: ```suggestion
Following are CLI tools that are useful during development:
```

.. code-block:: console

    cookiecutter # tool to bootstrap new modules from templates
    chromedriver # selenium driver for chrome (used for end-to-end testing)
    gh           # GitHub CLI client (useful e.g for checking out PRs)
    gifify       # make short screen recordings for bug reports
    git          # our version control system
    hub          # extends git with github features

General CLI tools:

.. code-block:: console

    htop            # a better top

>     tree            # pretty print a directory structur

Author: ```suggestion
    tree            # pretty print a directory structure
```

install ``pyenv`` and ``nvm`` - both tools manage version of python and node
respectively. Install the following packages:

.. code-block:: console

    nvm
    pyenv
    pyenv-virtualenv
    pyenv-virtualenvwrapper

Once you have installed above packages, you can proceed with installing Python
versions. The following will install Python 3.6, 3.7 and 3.8 and set the
default Python installation to Python 3.8 (node you can always install the
latest patch-level release):

.. code-block:: console

    $ pyenv install 3.6.9
    $ pyenv install 3.7.8
    $ pyenv install 3.8.5
    $ pyenv global 3.8.5

You should edit your `.bashrc` or `.zshrc` file to initialise pyenv:

.. code-block:: sh

    # nvm setup
    export NVM_DIR="$HOME/.nvm"
    [ -s "/usr/local/opt/nvm/nvm.sh" ] && . "/usr/local/opt/nvm/nvm.sh"

    # pyenv
    eval "$(pyenv init -)"

    # pyenv-virtualenv
    eval "$(pyenv virtualenv-init -)"

    # pyenv-virtualenvwrapper
    pyenv virtualenvwrapper

Now, you can create e.g. Python virtual environments using the following
commands:

.. code-block:: console

    $ mkvirtualenv <name>
    $ mkvirtualenv -p python3.7 <name>
    $ workon <name>

Linux
~~~~~

> Want to write it? Contact us on the chat!

Author: Anyone on Linux which can help with this part? 😄
